✨ feat : WebSocket 기반 실시간 메시지 전송 기능 및 권한 검증 로직 리팩터링

BE12-OZ opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #44 
- 작업 요약: WebSocket 기반 실시간 메시지 전송 기능 통합 및 권한 검증 로직 리팩터링

## 📄 상세 내용
- [x] WebSocket 기반 실시간 메시지 전송 및 DB 저장 기능
- 파일: apps/chat/consumers.py
	1.	ChatConsumer 정의
	•	channels.generic.websocket.AsyncWebsocketConsumer 상속.
	•	WebSocket 연결·해제 및 메시지 송수신 전 과정을 처리.
	
	2.	핵심 메서드
	•	connect():
	•	study_group_id를 URL 패턴에서 추출.
	•	channel_layer.group_add를 통해 그룹에 참여.
	•	권한 확인 로직 추가:
	         is_user_in_group(user, study_group_id)를 호출해 접속 유저가 스터디 그룹 멤버인지 확인.
	         멤버가 아닐 경우 await self.close()로 연결 즉시 거부.
	•	isinstance(user, User) 타입 가드 적용 → mypy 타입 안정성 확보.
	•	disconnect():
	•	연결 종료 시 channel_layer.group_discard로 그룹에서 제거.
	•	receive():
	•	클라이언트로부터 "chat.message" 이벤트 수신.
	•	인증된 사용자(scope["user"].is_authenticated)만 메시지 처리.
	•	@database_sync_to_async로 감싼 create_chat_message() 호출 → 비동기 환경에서도 안전하게 DB 저장.
	•	이후 channel_layer.group_send로 그룹 내 모든 사용자에게 실시간 브로드캐스트.
	•	chat_message():
	•	그룹으로부터 전달된 메시지를 클라이언트에게 전송.
	
	3.	DB Persistence 처리
	•	Django ORM은 동기 방식 → Channels의 비동기 환경에서 @database_sync_to_async로 감싸 안전하게 실행.
	
- [x] WebSocket 라우팅 및 설정
- 파일:
	•	apps/chat/routing.py
	•	엔드포인트: ws/study-groups/(?P<study_group_id>\d+)/chat/
	•	API 명세서의 엔드포인트와 일치하도록 ChatConsumer 매핑.
	•	config/asgi.py
	•	Channels 기반 Django 구성.
	•	ProtocolTypeRouter로 HTTP/WS 분리 처리.
	•	AuthMiddlewareStack 적용 → Django 인증 시스템을 WebSocket에서도 사용 가능.
	
- [x] 비동기 테스트 코드 작성
- 파일: apps/chat/tests/tests.py
	•	channels.testing.WebsocketCommunicator를 사용해 비동기 WebSocket 테스트 구현.
	•	시나리오:
	1.	가상 클라이언트가 메시지를 전송.
	2.	서버가 해당 메시지를 브로드캐스트하는지 검증.
	3.	ChatMessage.objects.aexists()로 실제 DB 저장 여부 확인.
	•	DB 저장 및 실시간 전송 기능 검증 완료.
                `test_group_member_can_connect_and_send_message` 테스트가 통과 
                 -> 'WebSocket 연결 및 송수신'과 'DB 저장 기능'이 정상 작동함을 증명
                 `test_non_group_member_is_rejected` 테스트가 통과
                 -> '비인가 사용자 연결 거부' 기능이 정상 작동함을 증명
                 
- [x] 권한 확인 로직 리팩터링 (Refactor)
- 커밋: refactor: 채팅 연결 권한 확인 로직 추가 및 테스트 보강
	1.	connect() 내 권한 확인 로직 추가
	•	이전에는 모든 사용자가 WebSocket 연결 가능했던 보안 취약점 존재.
	•	개선:
	•	is_user_in_group(user, study_group_id) 호출로 유저의 스터디 그룹 소속 여부 확인.
	•	비인가 사용자는 await self.close() 호출로 즉시 연결 거부.
	•	isinstance(user, User) 타입 가드 적용 → mypy에서 scope["user"] 타입 명확화.
	•	# type: ignore 최소화 → 정적 타입 안정성 강화.
	
	2.	테스트 코드 보강
	•	추가 테스트 케이스: test_non_group_member_is_rejected
	•	스터디 그룹 멤버가 아닌 사용자의 연결이 거부되는지 검증.
	•	communicator.connect() 결과로 connected == False를 확인.
	•	권한 검증 로직의 정상 동작 보장.
	
	3.	기타 수정 사항
	•	apps/chat/urls.py 내 불필요한 import 제거 → 경고 메시지 해결.

- [x] 프로젝트 구조 정리
- HTTP 기반 메시지 전송 API 제거
	•	WebSocket으로 전환.
	•	제거 항목:
	•	config/urls.py 내 apps.chat.urls.chat_urls 경로 제거.
	•	apps/chat/views/chat_views.py
	•	apps/chat/serializers/chat_serializers.py
	•	apps/chat/urls/chat_urls.py
	•	관련 디렉터리(views, serializers, urls) 정리 완료.

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항
- [x] 최종 확인 사항
        •	코드 포매팅(isort, black) 통과
	•	mypy 정적 타입 검사 통과
	•	WebSocket 연결 및 송수신 테스트 성공
	•	비인가 사용자 연결 거부 테스트 성공
	•	DB 저장 기능 정상 작동
	•	JWT 인증 연동 완료 시, 실시간 채팅 기능 작동 테스트 예정

## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

BE12-OZ committed:
♻️ refactor: 알림 및 채팅 기능 의존성/설정 통일

# Conflicts:
#	poetry.lock
#	pyproject.toml

# Conflicts:
#	config/settings/base.py
#	poetry.lock
#	pyproject.toml

BE12-OZ committed:
✨ feat: 실시간 채팅 메시지 전송 및 DB 저장 기능 구현

- `ChatConsumer`의 `receive` 메소드에 `ChatMessageService`를 호출하여 DB에 메시지를 저장하는 로직을 추가했습니다.
- 'database_sync_to_async' 데코레이터를 사용하여 비동기 환경에서 ORM 작업을 안전하게 처리했습니다.
- `channels.testing.WebsocketCommunicator`를 사용하여, 메시지 수신 시 DB에 정상적으로 저장되는지 검증하는 통합 테스트 코드를 추가했습니다.

BE12-OZ committed:
🎨 style: isort 포맷팅 규칙에 맞게 임포트 순서 수정

2joonkim commented on apps/chat/consumers.py:
```diff
@@ -0,0 +1,78 @@
+        self.room_group_name = f"chat_{self.study_group_id}"
+
+        # Join room group
+        await self.channel_layer.group_add(self.room_group_name, self.channel_name)
```
DB에서 해당 StudyGroup이 실제 존재하는지 검증하지 않아서 
이거 존재하지 않은 아이디로도 접속이 가능한거 아닌가요?! 
이 부분 확인한번 해보셨으면 좋겠습니다!

2joonkim commented on apps/chat/consumers.py:
```diff
@@ -0,0 +1,78 @@
+
+    # Receive message from WebSocket
+    async def receive(self, text_data: str | None = None, bytes_data: bytes | None = None) -> None:
+        if not text_data:
```
이 부분 잘못된json 포맷의 데이터가 들어 왔을 때 JSONDecodeError가 터질 수 있을 거 같은데
방어 코드가 필요해 보입니다.

2joonkim commented on apps/chat/consumers.py:
```diff
@@ -0,0 +1,78 @@
+        sender_id = event["sender_id"]
+
+        # Send message to WebSocket
+        await self.send(text_data=json.dumps({"type": "chat.message", "message": message, "sender_id": sender_id}))
```
저희 서비스는 한글이 주이기 때문에
```ensure_ascii=False```
한줄 추가하면 한글이 깨질일이 방지됩니다!! 이 부분은 확인해서 보완해주시면 좋을거같아요

BE12-OZ replied:
조교님 리뷰 감사드립니다. 
상기 말씀해주신 부분들 체크해서 수정보완 하겠습니다 !

Meoyoug commented on apps/chat/tests/tests.py:
추후에 채팅관련 API 에 대해서도 TestCode 를 작성해야하니 
filename 을 test_chat_consumers.py 로 변경부탁드립니당

Meoyoug commented on apps/chat/consumers.py:
```diff
@@ -0,0 +1,78 @@
+        if not text_data:
+            return
+
+        text_data_json = json.loads(text_data)
```
정확히는 여기서 loads 호출할 때 JSON 형식의 데이터가 아니라면 터질겁니다

Meoyoug commented on apps/chat/consumers.py:
컨슈머 중에 AsyncJsonWebsocketConsumer 를 사용하시면 Json의 형식 검증이나 Json 포매팅에 대한 염려를 줄일 수 있을 겁니다 찾아보시고 적용해보시면 좋겠어요

Meoyoug commented on apps/chat/tests/tests.py:
```diff
@@ -1,3 +1,71 @@
+        self.assertTrue(message_exists)
+
+        # 7. 웹소켓 연결 종료
+        await communicator.disconnect()
```
유저 2명이 웹소켓 접속을 하고 특정 채팅방에 접속했을 때를 가정해서 테스트를 작성해보시면 더 신뢰도가 높을 것 같아요 본인한테만 채팅메시지가 브로드캐스팅되는 것과 타인에게도 브로드캐스팅이 된다는 것은 차이가 크니까요

BE12-OZ replied:
코치님 리뷰 감사드립니다.
체크해주신 부분 더 찾아보고 수정 보완하도록 하겠습니다 !

BE12-OZ committed:
♻️ refactor: WebSocket 컨슈머 및 테스트 코드 리뷰 피드백 반영

- `AsyncJsonWebsocketConsumer`를 사용하도록 리팩토링하여 JSON 처리 로직 개선
- `connect` 시 `StudyGroup` 존재 및 그룹 멤버 여부를 확인하는 권한 검증 로직 추가
- `mypy` 타입 에러 해결을 위한 코드 수정 (import 경로, 타입 명시, ignore 처리)
- 두 명의 사용자가 참여하도록 테스트 케이스를 개선하여 브로드캐스팅 검증 강화
- `isort`, `black` 포맷팅 적용

Meoyoug commented on apps/chat/consumers.py:
```diff
@@ -0,0 +1,93 @@
+            return False
+        return bool(GroupMember.objects.filter(study_group_id=self.study_group_id, user=user).exists())
+
+    @database_sync_to_async  # type: ignore[misc]
```
이 데코레이터 대신 이제 
```
study_group = StudyGroup.objects.aget(id=self.study_group_id)
```

이렇게도 가능해요~!

Meoyoug replied:
함수를 async  비동기함수로 바꾸고요!

Meoyoug commented on apps/chat/consumers.py:
```diff
@@ -0,0 +1,93 @@
+    def is_group_member(self, user: AbstractBaseUser) -> bool:
+        if not user.is_authenticated or not isinstance(user, get_user_model()):
+            return False
+        return bool(GroupMember.objects.filter(study_group_id=self.study_group_id, user=user).exists())
```
이 부분들도 비동기함수 선언인 async def 로 선언하고

exists -> aexists 로 변경하시면 비동기 ORM 처리가 가능합니다!

https://docs.djangoproject.com/en/5.2/ref/models/querysets/

BE12-OZ replied:
넵 ! 리뷰 감사드립니다. 반영해서 수정 보완하도록 하겠습니다.

Meoyoug reviewed (changes requested):
웹소켓 컨슈머에서 비동기처리와 비동기 ORM 메서드를 사용해보면 좋을 것 같아요!

BE12-OZ committed:
✨ feat: 채팅 앱 비동기 ORM 피드백 반영 및 mypy 오류 수정

`apps/chat/consumers.py`에 코치님의 비동기 처리 및 비동기 ORM 메서드 피드백을 반영했습니다.
- `is_valid_study_group` 및 `is_group_member` 함수를 `async def`로 변경하고 `aexists()`를 활용했습니다.
- `create_chat_message`에서 `await StudyGroup.objects.aget()`을 사용하도록 했습니다.
`apps/chat/tests/test_chat_consumers.py`의 `mypy` 오류를 수정했습니다.
- `AsgiTestCase` 서브클래싱에 `type: ignore[misc]`를 추가했습니다.
- `_async_test_logic`에 `-> None` 반환 타입 주석을 추가했습니다.
- `asyncio`를 임포트했습니다.

BE12-OZ committed:
🚚 build: 누락된 channels/channels-redis 종속성 추가

BE12-OZ committed:
fix: consumers.py rebase 충돌 해결 및 기능 복구

develop 브랜치로 rebase 하는 과정에서 consumers.py 파일에 충돌이 발생하여 이를 해결했습니다.

rebase로 인해 누락되었던 WebSocket 연결 시 읽음 처리 로직을 다시 추가하여 복구했습니다.

또한, 충돌을 해결하면서 Django ORM의 비동기 메서드 호출 오류(aget_latest)를 올바른 방식인 order_by("-created_at").afirst()로 수정했습니다.

BE12-OZ committed:
fix: consumers.py rebase 충돌 해결 및 기능 복구

develop 브랜치로 rebase 하는 과정에서 consumers.py 파일에 충돌이 발생하여 이를 해결했습니다.

rebase로 인해 누락되었던 WebSocket 연결 시 읽음 처리 로직을 다시 추가하여 복구했습니다.

또한, 충돌을 해결하면서 Django ORM의 비동기 메서드 호출 오류(aget_latest)를 올바른 방식인 order_by("-created_at").afirst()로 수정했습니다.
